### ISSUE TITLE
calculateResolvedType cannot be stopped.

### ISSUE THREAD

**Issue by zhangziqiang1** [javaparser/javaparser#2990]
### I've had a very weird problem, and this doesn't seem to happen.
I had a strange problem and this shouldn't have happened, but it happened and I tried to call  `methodCallExpr.getScope().get().calculateResolvedType();
`
The program cannot be stopped.
Here's my case.

```
public void bad(String sql, Map<Integer, Object> condition) {
        Connection conn = null;
        PreparedStatement stmt = null;
        ResultSet rs = null;
        if (conn != null) {
            try {
                StringBuffer temSql = new StringBuffer(sql);
                stmt = conn.prepareStatement(temSql.toString());
                for (final Map.Entry<Integer, Object> entry : condition.entrySet()) {
                    stmt.setObject(entry.getKey(), entry.getValue());
                }
                rs = stmt.executeQuery();
                while (rs.next()) {
                    TestModel testModel = new TestModel();
                    testModel.setDeviceIp(rs.getString("test"));
                    testModel.setZoneId(rs.getLong("test"));
                    testModel.setZoneIp(rs.getString("test"));
                    testModel.setBizId(rs.getLong("test"));
                    testModel.setStartTimeAlert(rs.getLong("test"));
                    testModel.setStartTimeAttack(rs.getLong("test"));
                    testModel.setEndTime(rs.getLong("test"));
                    testModel.setDuration(rs.getLong("test"));
                    testModel.setDropKbits(rs.getLong("test"));
                    testModel.setDropPackets(rs.getLong("test"));
                    testModel.setAttackType(rs.getInt("test"));
                    testModel.setAttackStatus(rs.getInt("test"));
                    testModel.setProtocol(rs.getInt("test"));
                    testModel.setPort(rs.getInt("test"));
                    testModel.setLimit(rs.getLong("test"));
                    testModel.setPpsOrKbps(rs.getInt("test"));
                    testModel.setMaxDropKbps(rs.getLong("test"));
                    testModel.setMaxDropPps(rs.getLong("test"));
                    // methodCallExpr.getScope().get().calculateResolvedType();  The program is suspended and cannot be ended.
                    AttackFactory.getInstance();
                }
            } catch (SQLException ex) {
            }
        }

    }
```

```
public static void main(String[] args) {
        StaticJavaParser.getConfiguration().setLanguageLevel(ParserConfiguration.LanguageLevel.JAVA_9);
        CombinedTypeSolver combinedTypeSolver = new CombinedTypeSolver();
        combinedTypeSolver.add(new ReflectionTypeSolver());
        combinedTypeSolver.add(new JavaParserTypeSolver("D:\\JavaParserSample\\src\\main\\java"));
        StaticJavaParser.getConfiguration().setSymbolResolver(new JavaSymbolSolver(combinedTypeSolver));
        try {
            CompilationUnit node = StaticJavaParser.parse(new FileReader(new File("D:\\JavaParserSample\\src\\main\\java\\javaParser18\\Error1.java")));

            node.findAll(MethodDeclaration.class).forEach(methodDeclaration -> {
                if (methodDeclaration.getNameAsString().equals("bad")) {
                    methodDeclaration.findAll(MethodCallExpr.class).forEach(methodCallExpr -> {
                        if (methodCallExpr.toString().equals("AttackFactory.getInstance()")) {
                            // The program cannot be stopped.
                            methodCallExpr.getScope().get().calculateResolvedType();
                        }
                    });
                }

            });

        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
    }
```
[JavaParserSample.zip](https://github.com/javaparser/javaparser/files/5729907/JavaParserSample.zip)

**Comment by zhangziqiang1** [comment]
@jlerbsc Thank you for taking the time to look at this question. I have provided a test case and I use version 3.18.0.

**Comment by jlerbsc** [comment]
Hi @zhangziqiang1 i think a quick workaround is to use the qualified name for AttackFactory.

**Comment by jlerbsc** [comment]
I fact it's not a bug but an issue related to performance.

**Comment by zhangziqiang1** [comment]
That's not a problem. I hope you can solve this performance problem. It doesn't seem to happen here.

**Comment by jlerbsc** [comment]
I'm working on it. " It doesn't seem to happen here.Y you mean with the proposed workraround?

**Comment by zhangziqiang1** [comment]
I'm glad to see you're dealing with this. Expect a new version

**Comment by zhangziqiang1** [comment]
I'm glad to see you're dealing with this. Expect a new version

**Comment by jlerbsc** [comment]
I tried to implement a thread local cache with dynamic proxy decorator (and Cacheable method anotation) on StatementContext but I failed to pass the slow tests :-( . So I deleted this PR.

**Comment by zhangziqiang1** [comment]
This problem has been resolved in 3.20.0.
